controllers/templatesync/template_sync.go

```diff
@@ -866,3 +1072,84 @@ func getLatestStatusMessage(pol *policiesv1.Policy, tIndex int) string {
 
 	return tmplDetails.History[0].Message
 }
+
+func hasClusterwideFinalizer(pol *policiesv1.Policy) bool {
+	for _, finalizer := range pol.Finalizers {
+		if finalizer == utils.ClusterwideFinalizer {
+			return true
+		}
+	}
+
+	return false
+}
+
+// removeFinalizer iterates over the finalizers and removes the one specified from the policy.
+func removeFinalizer(pol *policiesv1.Policy, finalizer string) {
+	i := 0
+	finalizersLength := len(pol.Finalizers)
+
+	for i < finalizersLength {
+		if pol.Finalizers[i] == finalizer {
+			pol.Finalizers = append(pol.Finalizers[:i], pol.Finalizers[i+1:]...)
+			finalizersLength--
+		} else {
+			i++
+		}
+	}
+}
+
+// finalizerCleanup handles any steps required for cleaning up objects on the cluster prior to
+// removing the policy finalizer.
+func finalizerCleanup(
+	ctx context.Context, pol *policiesv1.Policy, rMapper meta.RESTMapper, dClient dynamic.Interface,
+) error {
+	var errorList []error
+
+	for _, policyT := range pol.Spec.PolicyTemplates {
+		object, gvk, err := unstructured.UnstructuredJSONScheme.Decode(policyT.ObjectDefinition.Raw, nil, nil)
+		if err != nil {
+			errorList = append(errorList, fmt.Errorf("failed to decode policy template with error: %w", err))
+
+			policyUserErrorsCounter.WithLabelValues(pol.Name, "", "format-error").Inc()
+
+			continue
+		}
+
+		// Skip template if the name isn't found
+		var tName string
+		if tMetaObj, ok := object.(metav1.Object); ok {
+			tName = tMetaObj.GetName()
+		} else {
+			continue
+		}
+
+		// If there was an error getting the mapping, it's likely because the CRD is gone--skip this
+		// template since Kubernetes garbage collection will take care of it
+		mapping, err := rMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			continue
+		}
+
+		// Instantiate a dynamic client
+		rsrc := mapping.Resource
+		res := dClient.Resource(rsrc)
+
+		// Delete clusterwide objects
+		if mapping.Scope == meta.RESTScopeRoot {
+			// Delete object, ignoring not found errors
+			err := res.Delete(ctx, tName, metav1.DeleteOptions{})
+			if err != nil && !k8serrors.IsNotFound(err) {
+				policySystemErrorsCounter.WithLabelValues(pol.Name, tName, "delete-error").Inc()
+
+				errorList = append(errorList, fmt.Errorf("failed to delete "+gvk.Kind+" with error: %w", err))
+			}
+		}
+	}
+
+	var err error
+	for _, errorItem := range errorList {
```

Comment:
Optional, but this is a pretty cool way to aggregate errors:
https://pkg.go.dev/k8s.io/apimachinery@v0.23.9/pkg/util/errors#Aggregate

Reply:
Oh nice--I saw things coming in Go 1.20, but I'll poke at this instead!

Reply:
Having multi errors in go 1.20 will be nice